spring-cloud-sleuth-autoconfigure/src/test/java/org/springframework/cloud/sleuth/autoconfig/wavefront/WavefrontTracingIntegrationTests.java

```diff
@@ -0,0 +1,392 @@
+import brave.sampler.Sampler;
+import com.wavefront.sdk.common.Pair;
+import com.wavefront.sdk.common.WavefrontSender;
+import com.wavefront.sdk.common.application.ApplicationTags;
+import com.wavefront.sdk.entities.histograms.HistogramGranularity;
+import com.wavefront.sdk.entities.tracing.SpanLog;
+import io.opentracing.Tracer;
+import io.opentracing.tag.Tags;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.Test;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.test.autoconfigure.actuate.metrics.AutoConfigureMetrics;
+import org.springframework.boot.test.autoconfigure.web.reactive.AutoConfigureWebTestClient;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Primary;
+import org.springframework.core.env.Environment;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Controller;
+import org.springframework.test.annotation.DirtiesContext;
+import org.springframework.test.web.reactive.server.WebTestClient;
+import org.springframework.util.StringUtils;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * Integration tests for tracing.
+ */
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT,
+		classes = WavefrontTracingIntegrationTests.Config.class,
+		properties = { "wavefront.application.name=IntegratedTracingTests", "spring.application.name=test_service" })
+@AutoConfigureWebTestClient
+@AutoConfigureMetrics
+@DirtiesContext
+public class WavefrontTracingIntegrationTests {
+
+	@Autowired
+	private WebTestClient client;
+
+	@Autowired
+	private BlockingDeque<SpanRecord> spanRecordQueue;
+
+	@Test
+	void sendsToWavefront() {
+		this.client.get().uri("/api/fn/10").header("b3", "0000000000000001-0000000000000003-1-0000000000000002")
+				.exchange().expectStatus().isOk();
+
+		SpanRecord spanRecord = takeRecord(spanRecordQueue);
+		assertThat(spanRecord.traceId).hasToString("00000000-0000-0000-0000-000000000001");
+		assertThat(spanRecord.parents).extracting(UUID::toString)
+				.containsExactly("00000000-0000-0000-0000-000000000003");
+		assertThat(spanRecord.followsFrom).isNull();
+		// This tests that RPC spans do not share the same span ID
+		assertThat(spanRecord.spanId.toString()).isNotEqualTo("00000000-0000-0000-0000-000000000003")
+				.matches("^[0-9a-f]{8}\\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\\b[0-9a-f]{12}$");
+		assertThat(spanRecord.name).isEqualTo("GET /api/fn/{id}");
+
+		// spot check the unit is valid (millis not micros)
+		long currentTime = System.currentTimeMillis();
```

Comment:
Isn't the `startMillis` come from `nanoTime()` originally? Is so, would not it be better to get the `currentTime` from there too?
```suggestion
		long currentTime = TimeUnit.NANOSECONDS.toMillis(System.nanoTime());
```